WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,243 @@
+			{
+				throw new ArgumentException("Last event sequenceId is not a positive integer.", nameof(wrappedEvents));
+			}
+
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("Event sequence ids are inconsistent.", nameof(wrappedEvents));
+			}
+
+			var aggregateEventsBatches = AggregatesEventsBatches.GetOrAdd(aggregateType, _ => new());
+			var (tailSequenceId, events) = aggregateEventsBatches.GetOrAdd(
+				aggregateId,
+				_ => (0, ImmutableList<WrappedEvent>.Empty));
+
+			if (tailSequenceId + 1 < firstSequenceId)
+			{
+				throw new ArgumentException($"Invalid firstSequenceId (gap in sequence ids) expected: '{tailSequenceId + 1}' given: '{firstSequenceId}'.", nameof(wrappedEvents));
+			}
+
+			// no action
+			Validated();
+
+			var newEvents = events.AddRange(wrappedEventsList);
+
+			// Atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches.
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, newEvents),
+				comparisonValue: (firstSequenceId - 1, events)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException(
+					$"Conflict while committing events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
+			}
+			Appended(); // no action
+
+			// If it is a first event for given aggregate.
+			if (tailSequenceId == 0)
+			{
+				// Add index of aggregate id into the dictionary.
+				IndexNewAggregateId(aggregateType, aggregateId);
+			}
+			return Task.CompletedTask;
+		}
+
+		/// <inheritdoc/>
+		public Task<IReadOnlyList<WrappedEvent>> ListEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			long afterSequenceId = 0,
+			int? maxCount = null)
+		{
+			if (AggregatesEventsBatches.TryGetValue(aggregateType, out var aggregateEventsBatches) &&
+				aggregateEventsBatches.TryGetValue(aggregateId, out var value))
+			{
+				var result = value.Events;
+
+				if (afterSequenceId > 0)
+				{
+					var foundIndex = result.BinarySearch(
+						new WrappedEvent(afterSequenceId),
+						WrappedEventSequenceIdComparer);
+					if (foundIndex < 0)
+					{
+						// Note: this is because of BinarySearch() documented implementation
+						// returns "bitwise complement"
+						// see: https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1.binarysearch
+						// The zero-based index of item in the sorted List, if item is found;
+						// otherwise, a negative number that is the bitwise complement
+						// of the index of the next element that is larger than item or,
+						// if there is no larger element, the bitwise complement of Count.
+						foundIndex = ~foundIndex;
+					}
+					else
+					{
+						foundIndex++;
+					}
+					result = result.GetRange(foundIndex, result.Count - foundIndex);
+				}
+
+				if (maxCount < result.Count)
+				{
+					result = result.GetRange(0, maxCount.Value);
+				}
```

Comment:
After closing brace, there is a blank line. Similarly, `if` is preceded by a blank line. 

You can have a look to other files to find out if your code style matches the rest of code base.

Reply:
sorry, but it is not documented here https://github.com/zkSNACKs/WalletWasabi/blob/master/CONTRIBUTING.md . also if you want such tedious rules just configure it to codemaid or .editorconfig to be done automatically. I'm not doing this micromanagement of whitespace.

Reply:
Hard rules that block merging need to be explicitly defined and documented and ideally automated. And there needs to be as little of them as possible. It is hard to remember more than 1 A4 page. For you personal esthetical preferences I don't really care because different people have different preferences anyway.

Reply:
Ok, it's blatantly clear. I won't suggest these for *your* PRs anymore.

Reply:
where is complete list of merge blocking code style rules? 🧐

Reply:
kiminuo himself told that all of his recommendations are not binding and I can ignore and he will stop making them. This is all very confusing.

I there are hard rules I wanna know them. Explicitly. If the recommendation is soft I wanna know it. Otherwise is all very confusing. I'm not technically able to follow unknown rules, sorry.

Reply:
I don't even know what is this comment about. Where should I put the new line? Can't this be automated? Inserting new lines based on algorithmic rules is job for CodeMaid 🤷‍♂️

Reply:
Is this commit ddd547a7e9e899ca6f28693e1dc527f535cb4c39  what you want?

Reply:
> kiminuo himself told that all of his recommendations are not binding and I can ignore and he will stop making them. This is all very confusing.

I said that because it was clear you would disregard them anyway.

> I don't even know what is this comment about. Where should I put the new line? Can't this be automated? Inserting new lines based on algorithmic rules is job for CodeMaid 🤷‍♂️

The code style is:

```cs
// blank line
if (something) 
{
}
// blank line
// next command
```

> Is this commit ddd547a what you want?

Yes.